File: csharp/MetaObjects/DataConverter.cs
// DataConverter — convert a raw JSON value to a valid AttrValue.
//
// Ported 1:1 from typescript/packages/metadata/src/data-converter.ts.
//
// AttrValue contract (locked):
//   C# stores attr values as object? constrained at runtime to exactly:
//     - string
//     - long   (all integers; never int)
//     - double (non-integer numbers)
//     - bool
//     - IReadOnlyList<string>
//   This single-path number model keeps the canonical serializer correct.
//   Callers that receive object? MUST test type before use; casting to int
//   or float is forbidden.

using System.Text.Json;
using System.Text.RegularExpressions;

namespace MetaObjects;

/// <summary>
/// Convert a raw parsed JSON value to a valid AttrValue for a known DataType.
/// </summary>
/// <remarks>
/// Mirrors <c>data-converter.ts</c>: two exported functions, <c>convertToDataType</c>
/// (type-directed, used by the parser for declared attrs) and <c>toAttrValue</c>
/// (no known type, used for undeclared @ attrs).
/// </remarks>
public static class DataConverter
{
    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /// <summary>
    /// Convert a raw JSON element to a valid AttrValue for a <em>known</em>
    /// <see cref="DataType"/>. Throws <see cref="FormatException"/> for values
    /// that cannot become a valid attr value (null, nested arrays, objects that
    /// cannot be stringified meaningfully). Arrays always return
    /// <see cref="IReadOnlyList{string}"/> regardless of the scalar DataType.
    /// </summary>
    /// <remarks>
    /// Mirrors TS <c>convertToDataType(dataType, raw)</c>.
    /// </remarks>
    public static object ToAttrValue(JsonElement element) =>
        ToAttrValueInternal(element, fnName: "toAttrValue", typeDirected: false, dataType: null);

    /// <summary>
    /// Convert a raw JSON element to a valid AttrValue, coercing toward the
    /// given <see cref="DataType"/>. For declared <c>@</c>-attrs whose type is
    /// known from the registered schema. Arrays always return
    /// <see cref="IReadOnlyList{string}"/> regardless of the scalar DataType;
    /// scalars are coerced (e.g. string "42" → long 42, string "true" → bool).
    /// </summary>
    /// <remarks>
    /// Mirrors TS <c>convertToDataType(dataType, raw)</c>.
    /// </remarks>
    public static object ConvertToDataType(DataType dataType, JsonElement element) =>
        ToAttrValueInternal(element, fnName: "convertToDataType", typeDirected: true, dataType: dataType);

    /// <summary>
    /// Convert an already-parsed attr value (i.e. <c>object?</c> from <see cref="MetaData.OwnAttr"/>)
    /// to the given <see cref="DataType"/>. Used by <c>MetaField.DefaultValue()</c> to re-coerce
    /// the <c>@default</c> attr that was already stored (analogous to TS
    /// <c>convertToDataType(dataType, raw)</c> where <c>raw</c> is an already-parsed value).
    /// </summary>
    public static object? ConvertToDataType(DataType dataType, object? raw)
    {
        if (raw is null) return null;

        // Array → IReadOnlyList<string> (passthrough — already converted at parse time).
        if (raw is IReadOnlyList<string> list) return list;

        // Type-directed coercion on scalar values.
        return dataType switch
        {
            DataType.Boolean => raw is bool b ? (object)b
                             : raw is string s && s == "true" ? true
                             : raw is string s2 && s2 == "false" ? false
                             : raw,
            DataType.Int or DataType.Long =>
                raw is long ? raw
                : raw is double d && d == Math.Floor(d) ? (object)(long)d
                : raw is string str && long.TryParse(str, out long parsed) ? parsed
                : raw,
            DataType.Double =>
                raw is double ? raw
                : raw is long lg ? (object)(double)lg
                : raw is string str2 && double.TryParse(str2,
                    System.Globalization.NumberStyles.Float,
                    System.Globalization.CultureInfo.InvariantCulture, out double pd) ? pd
                : raw,
            // String, Object, Date — string form (mirrors TS `default: String(raw)`).
            _ => raw is string ? raw : raw.ToString() ?? raw,
        };
    }

    // -------------------------------------------------------------------------
    // Internal dispatch
    // -------------------------------------------------------------------------

    private static object ToAttrValueInternal(
        JsonElement element,
        string fnName,
        bool typeDirected,
        DataType? dataType)
    {
        // Reject null — mirrors rejectNullish in TS.
        if (element.ValueKind == JsonValueKind.Null || element.ValueKind == JsonValueKind.Undefined)
            throw new FormatException($"{fnName}: null is not a valid attr value");

        // Arrays → IReadOnlyList<string> always (regardless of scalar DataType).
        if (element.ValueKind == JsonValueKind.Array)
            return ToStringArray(element, fnName);

        if (!typeDirected)
        {
            // toAttrValue — no type-directed coercion; return the JSON scalar as-is.
            return element.ValueKind switch
            {
                JsonValueKind.String  => element.GetString()!,
                JsonValueKind.True    => (object)true,
                JsonValueKind.False   => (object)false,
                JsonValueKind.Number  => ParseNumber(element),
                _ => throw new FormatException(
                    $"toAttrValue: {element.ValueKind} is not a valid attr value"),
            };
        }

        // convertToDataType — type-directed coercion.
        return dataType! switch
        {
            DataType.Boolean => ToBoolean(element, fnName),
            DataType.Int     => ToInteger(element, fnName),
            DataType.Long    => ToInteger(element, fnName),
            DataType.Double  => ToDouble(element, fnName),
            // String, Object, Date — string form (mirrors TS `default: String(raw)`).
            _ => element.ValueKind == JsonValueKind.String
                    ? element.GetString()!
                    : element.GetRawText(),
        };
    }

    // -------------------------------------------------------------------------
    // Type coercions — each mirrors the matching TS helper
    // -------------------------------------------------------------------------

    /// <summary>
    /// Mirrors TS <c>toBoolean(raw)</c>:
    /// <list type="bullet">
    ///   <item>bool JSON → bool</item>
    ///   <item>string "true" → true, "false" → false</item>
    ///   <item>anything else → pass through as-is (leave for attr-schema validation)</item>
    /// </list>
    /// </summary>
    private static object ToBoolean(JsonElement element, string fnName)
    {
        if (element.ValueKind == JsonValueKind.True)  return true;
        if (element.ValueKind == JsonValueKind.False) return false;
        if (element.ValueKind == JsonValueKind.String)
        {
            string s = element.GetString()!;
            if (s == "true")  return true;
            if (s == "false") return false;
            // Not boolean-shaped — pass through as-is (TS does the same).
            return s;
        }
        // Non-boolean, non-string — pass through as raw text (best-effort, for attr-schema validation).
        return element.GetRawText();
    }

    /// <summary>
    /// Mirrors TS <c>toInteger(raw)</c>:
    /// <list type="bullet">
    ///   <item>integer JSON number → long (if in safe range)</item>
    ///   <item>unsafe-integer JSON number → string (verbatim, to preserve precision)</item>
    ///   <item>strict integer string (e.g. "42", "-7") → long (if safe)</item>
    ///   <item>float-notation whole-number string (e.g. "3.0") → long (if safe integer value)</item>
    ///   <item>anything else → pass through as-is</item>
    /// </list>
    /// </summary>
    private static object ToInteger(JsonElement element, string fnName)
    {
        // JS safe-integer range: [-(2^53 - 1), 2^53 - 1]
        const long JsSafeIntegerMax =  9007199254740991L;
        const long JsSafeIntegerMin = -9007199254740991L;

        if (element.ValueKind == JsonValueKind.Number)
        {
            if (element.TryGetInt64(out long lng))
                return lng;
            // Non-integer JSON number (e.g. 3.7) — pass through as double,
            // matching TS `toInteger` which returns `raw` (the number) unchanged.
            return element.GetDouble();
        }
        if (element.ValueKind == JsonValueKind.String)
        {
            string s = element.GetString()!;
            // Strict integer string (e.g. "42", "-7").
            if (Regex.IsMatch(s, @"^-?\d+$"))
            {
                if (long.TryParse(s, out long parsed))
                    return parsed;
                // Beyond range — keep verbatim.
                return s;
            }
            // Float-notation strings that represent a whole number (e.g. "3.0", "12.0").
            // Mirrors TS: convert only when Number.isSafeInteger(parsed) — values beyond
            // 2^53−1 would have already lost precision in the double; keep original string.
            if (double.TryParse(s, System.Globalization.NumberStyles.Float,
                    System.Globalization.CultureInfo.InvariantCulture, out double parsed2)
                && double.IsFinite(parsed2)
                && parsed2 == Math.Floor(parsed2))
            {
                long asLong = (long)parsed2;
                if (asLong >= JsSafeIntegerMin && asLong <= JsSafeIntegerMax)
                    return asLong;
                // Outside safe-integer range — keep original string verbatim (mirrors TS).
                return s;
            }
            // Not integer-shaped — pass through.
            return s;
        }
        // Non-numeric — pass through as raw text.
        return element.GetRawText();
    }

    /// <summary>
    /// Mirrors TS <c>toDouble(raw)</c>:
    /// <list type="bullet">
    ///   <item>any JSON number → double</item>
    ///   <item>finite numeric string → double</item>
    ///   <item>anything else → pass through as-is</item>
    /// </list>
    /// </summary>
    private static object ToDouble(JsonElement element, string fnName)
    {
        if (element.ValueKind == JsonValueKind.Number)
        {
            return element.GetDouble();
        }
        if (element.ValueKind == JsonValueKind.String)
        {
            string s = element.GetString()!;
            if (s.Trim() != "" && double.TryParse(s, System.Globalization.NumberStyles.Float,
                    System.Globalization.CultureInfo.InvariantCulture, out double parsed)
                && double.IsFinite(parsed))
            {
                return parsed;
            }
            return s;
        }
        return element.GetRawText();
    }

    /// <summary>
    /// Parse a JSON number element into <see cref="long"/> or <see cref="double"/>.
    /// Integers are always long; non-integers are double.
    /// </summary>
    private static object ParseNumber(JsonElement element)
    {
        if (element.TryGetInt64(out long lng))
            return lng;
        return element.GetDouble();
    }

    /// <summary>
    /// Mirrors TS <c>toStringArray(arr)</c>:
    /// every element is stringified; nested arrays and null elements throw.
    /// </summary>
    private static IReadOnlyList<string> ToStringArray(JsonElement element, string fnName)
    {
        var list = new List<string>(element.GetArrayLength());
        int i = 0;
        foreach (JsonElement el in element.EnumerateArray())
        {
            if (el.ValueKind == JsonValueKind.Array)
                throw new FormatException(
                    $"array element at index {i} is a nested array — not supported");
            if (el.ValueKind == JsonValueKind.Null || el.ValueKind == JsonValueKind.Undefined)
                throw new FormatException(
                    $"array element at index {i} is null/undefined");
            // String(el) equivalent: use the raw string value for strings, raw text otherwise.
            list.Add(el.ValueKind == JsonValueKind.String
                ? el.GetString()!
                : el.GetRawText());
            i++;
        }
        return list.AsReadOnly();
    }
}

File: server/typescript/packages/metadata/src/data-converter.ts
// Data converter — convert a raw parsed value to a *known* DataType.
//
// The TypeScript counterpart of Java's DataConverter.toTypeSafe: directional
// (toward a type already known from the registered schema), never inferential.
// Replaces the deleted value-coerce.ts. See
// docs/superpowers/specs/2026-05-17-data-converter-design.md.

import type { AttrValue } from "./shared/meta-data.js";
import {
  type DataType,
  DATA_TYPE_BOOLEAN,
  DATA_TYPE_INT,
  DATA_TYPE_LONG,
  DATA_TYPE_DOUBLE,
  DATA_TYPE_OBJECT,
} from "./data-type.js";

/**
 * Convert a raw parsed value to a valid AttrValue for a *known* DataType.
 * Throws for input that cannot be a valid attr value (null / undefined /
 * nested array). A plain object reaching a scalar branch is stringified by
 * `String()` — callers that must reject objects use `toAttrValue`.
 */
export function convertToDataType(dataType: DataType, raw: unknown): AttrValue {
  rejectNullish(raw, "convertToDataType");

  // An array value → string[] regardless of the scalar DataType — the only
  // array shape AttrValue admits (e.g. a stringArray attr).
  if (Array.isArray(raw)) return toStringArray(raw);

  switch (dataType) {
    case DATA_TYPE_BOOLEAN:
      return toBoolean(raw);
    case DATA_TYPE_INT:
    case DATA_TYPE_LONG:
      return toInteger(raw);
    case DATA_TYPE_DOUBLE:
      return toDouble(raw);
    case DATA_TYPE_OBJECT:
      // Object-typed attrs (filter / properties) store the object verbatim.
      // A non-object (e.g. a legacy JSON string) is returned as-is so the
      // attr-schema pass can reject it with ERR_BAD_ATTR_VALUE.
      return (typeof raw === "object" && raw !== null)
        ? (raw as AttrValue)
        : String(raw);
    // string, date (unreachable for attrs) — string form
    default:
      return String(raw);
  }
}

/**
 * Make a raw value a valid AttrValue with NO known type — for an undeclared
 * `@`-attr. No type-directed conversion and no inference: scalars are stored
 * as-is (a numeric-looking string stays a string), arrays become string[],
 * objects/null/undefined are rejected.
 */
export function toAttrValue(raw: unknown): AttrValue {
  rejectNullish(raw, "toAttrValue");
  if (Array.isArray(raw)) return toStringArray(raw);
  if (typeof raw === "string" || typeof raw === "number" || typeof raw === "boolean") {
    return raw;
  }
  throw new Error(`toAttrValue: ${typeof raw} is not a valid attr value`);
}

// --- helpers ---------------------------------------------------------------

function rejectNullish(raw: unknown, fn: string): void {
  if (raw === null) throw new Error(`${fn}: null is not a valid attr value`);
  if (raw === undefined) throw new Error(`${fn}: undefined is not a valid attr value`);
}

function toBoolean(raw: unknown): AttrValue {
  if (typeof raw === "boolean") return raw;
  if (raw === "true") return true;
  if (raw === "false") return false;
  // Not boolean-shaped — best-effort: leave it for attr-schema validation.
  return raw as AttrValue;
}

function toInteger(raw: unknown): AttrValue {
  if (typeof raw === "number") {
    if (Number.isInteger(raw) && !Number.isSafeInteger(raw)) return String(raw);
    return raw;
  }
  if (typeof raw === "string") {
    // Strict integer string (e.g. "42", "-7").
    if (/^-?\d+$/.test(raw)) {
      const parsed = Number(raw);
      // Beyond MAX_SAFE_INTEGER → keep verbatim to preserve precision.
      return Number.isSafeInteger(parsed) ? parsed : raw;
    }
    // Float-format strings that represent a whole number (e.g. "3.0", "12.0").
    // Metadata files sometimes store int attr values in float notation; convert
    // to the integer value to preserve byte-identical behavior with the old
    // coerceAttrValue, which parsed these via the double pattern.
    const parsed = Number(raw);
    if (Number.isFinite(parsed) && Number.isInteger(parsed) && Number.isSafeInteger(parsed)) {
      return parsed;
    }
  }
  return raw as AttrValue;
}

function toDouble(raw: unknown): AttrValue {
  if (typeof raw === "number") return raw;
  if (typeof raw === "string") {
    const parsed = Number(raw);
    if (raw.trim() !== "" && Number.isFinite(parsed)) return parsed;
  }
  return raw as AttrValue;
}

function toStringArray(arr: unknown[]): string[] {
  return arr.map((el, i) => {
    if (Array.isArray(el)) {
      throw new Error(`array element at index ${i} is a nested array — not supported`);
    }
    if (el === null || el === undefined) {
      throw new Error(`array element at index ${i} is null/undefined`);
    }
    return String(el);
  });
}
